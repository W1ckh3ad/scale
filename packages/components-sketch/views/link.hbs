{{!--
Scale https://github.com/telekom/scale

Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG

This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, You can obtain one at https://mozilla.org/MPL/2.0/.
--}}

{{#* inline "icon"}}m12.65 3 1.5 1.5h-9.65v13.5c0 .85.65 1.5 1.5 1.5h12c.85 0 1.5-.65 1.5-1.5v-8.15l1.5 1.5v6.65c0 1.65-1.35 3-3 3h-12c-1.65 0-3-1.35-3-3v-15zm10.05-1.65-1.65 7.9-2.6-2.6-7.55 7.55c-.15.15-.35.2-.55.2s-.4-.05-.55-.2c-.3-.3-.3-.75 0-1.05l7.55-7.55-2.6-2.6 7.95-1.65z{{/inline}}

<div data-sketch-artboard="Link" class="artboard">
    <h1 class="title--artboard">Link</h1>
    
    <h2 class="title--artboard-section">Link</h2>
    <div class="agent-states--grid" data-category="Not Visited">
        <div class="agent-states--item">
            <p class="agent-states--label">:enabled</p>
            <scale-link data-sketch-key="link-0" href="#">Link Text</scale-link>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:hover</p>
            <scale-link data-sketch-key="link-1" href="#" data-sketch-state="hover">Link Text</scale-link>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:focus</p>
            <scale-link data-sketch-key="link-2" href="#" data-sketch-state="focus">Link Text</scale-link>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:active/:selected</p>
            <scale-link data-sketch-key="link-3" href="#" data-sketch-state="active">Link Text</scale-link>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:disabled</p>
            <scale-link data-sketch-key="link-4" href="#" disabled="true">Link Text</scale-link>
        </div>
    </div>
    
    <h2 class="title--artboard-section">Link :visited</h2>
    <div class="agent-states--grid" data-category="Visited">
        <div class="agent-states--item">
            <p class="agent-states--label">:enabled</p>
            <scale-link data-sketch-key="link-5" href="/link">Link Text</scale-link>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:hover</p>
            <scale-link data-sketch-key="link-6" href="/link" data-sketch-state="hover">Link Text</scale-link>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:focus</p>
            <scale-link data-sketch-key="link-7" href="/link" data-sketch-state="focus">Link Text</scale-link>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:active/:selected</p>
            <scale-link data-sketch-key="link-8" href="/link" data-sketch-state="active">Link Text</scale-link>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:disabled</p>
            <scale-link data-sketch-key="link-9" href="/link" disabled="true">Link Text</scale-link>
        </div>
    </div>

    <h2 class="title--artboard-section">Link with icon</h2>
    <div class="agent-states--grid" data-category="Not Visited">
        <div class="agent-states--item">
            <p class="agent-states--label">:enabled</p>
            <scale-link data-sketch-key="link-10" icon="{{>icon}}" href="#">Link Text</scale-link>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:hover</p>
            <scale-link data-sketch-key="link-11" icon="{{>icon}}" href="#" data-sketch-state="hover">Link Text</scale-link>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:focus</p>
            <scale-link data-sketch-key="link-12" icon="{{>icon}}" href="#" data-sketch-state="focus">Link Text</scale-link>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:active/:selected</p>
            <scale-link data-sketch-key="link-13" icon="{{>icon}}" href="#" data-sketch-state="active">Link Text</scale-link>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:disabled</p>
            <scale-link data-sketch-key="link-14" icon="{{>icon}}" href="#" disabled="true">Link Text</scale-link>
        </div>
    </div>

    <h2 class="title--artboard-section">Link with icon :visited</h2>
    <div class="agent-states--grid" data-category="Visited">
        <div class="agent-states--item">
            <p class="agent-states--label">:enabled</p>
            <scale-link data-sketch-key="link-15" icon="{{>icon}}" href="/link">Link Text</scale-link>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:hover</p>
            <scale-link data-sketch-key="link-16" icon="{{>icon}}" href="/link" data-sketch-state="hover">Link Text</scale-link>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:focus</p>
            <scale-link data-sketch-key="link-17" icon="{{>icon}}" href="/link" data-sketch-state="focus">Link Text</scale-link>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:active/:selected</p>
            <scale-link data-sketch-key="link-18" icon="{{>icon}}" href="/link" data-sketch-state="active">Link Text</scale-link>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:disabled</p>
            <scale-link data-sketch-key="link-19" icon="{{>icon}}" href="/link" disabled="true">Link Text</scale-link>
        </div>
    </div>
</div>

<script src="/symbol_names.js"></script>
<script>
    Array.from(document.querySelectorAll("scale-link")).forEach((btn,i) => {
        btn.dataset.sketchSymbol = `Link / ${
                order(1, getCategoryName(btn))
            } / ${
                order(2, btn.getAttribute('icon') ? 'With Icon' : 'Without Icon')
            } / ${
                order(4, getStateName(btn))
            }`;
    });
</script>