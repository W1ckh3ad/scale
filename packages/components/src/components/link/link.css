.link {
  cursor: pointer;
  outline: none;
  text-decoration: none;
  font-weight: var(--scl-font-weight-medium);
  color: var(--scl-color-text-link);
  border-bottom: var(--scl-spacing-1) solid currentColor;
}
.link:hover.link:not(.link--disabled) {
  color: var(--scl-color-text-link-hover);
}
.link:focus:not(.link--disabled) {
  border-bottom: var(--scl-spacing-2) solid currentColor;
}
.link:active.link:not(.link--disabled) {
  color: var(--scl-color-text-link-active);
}
.link:visited.link:not(.link--disabled) {
  color: var(--scl-color-text-link-visited);
}
.link__wrapper {
  display: inline-flex;
  align-items: center;
}
.link__wrapper > *:not(:first-child) {
  margin-inline-start: var(--scl-spacing-4);
}
.link--block {
  display: block;
}
.link--disabled {
  cursor: not-allowed;
  color: var(--scl-color-grey-50);
}
