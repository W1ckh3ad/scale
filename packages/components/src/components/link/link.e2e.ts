import { newE2EPage } from '@stencil/core/testing';

describe('scale-link', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<scale-link>default</scale-link>');
    const element = await page.find('scale-link');
    expect(element).toHaveClass('hydrated');
  });
});
