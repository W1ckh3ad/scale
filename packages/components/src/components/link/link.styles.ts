import { JssStyle } from 'jss';

export const styles: JssStyle = {
  link: {
    textDecoration: 'none',
    outline: 'none',
    cursor: 'pointer',
    color: ({ color }) => color.link,
    '&:hover&:not($link--disabled)': {
      color: ({ color }) => color.link_hover,
      borderBottom: ({ size }) => `${size.underline}px solid currentColor`,
    },
    '&:focus:not($link--disabled)': {
      borderBottom: ({ size }) => `${size.border_focus}px solid currentColor`,
    },
    '&:active&:not($link--disabled)': {
      color: ({ color }) => color.link_active,
    },
    '&:visited&:not($link--disabled)': {
      color: ({ color }) => color.link_visited,
    },
  },
  link__wrapper: {
    display: 'inline-flex',
    alignItems: 'center',
    '& > *:not(:first-child)': {
      marginInlineStart: ({ spacing }) => spacing.char,
    },
  },
  'link--block': {
    display: 'block',
  },
  'link--disabled': {
    cursor: 'not-allowed',
    color: ({ color }) => color.disabled,
  },
};
