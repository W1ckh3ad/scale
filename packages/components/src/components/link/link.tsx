import { Component, h, Prop, Host } from '@stencil/core';
import classNames from 'classnames';

@Component({
  tag: 'scale-link',
  styleUrl: './link.css',
  shadow: true,
})
export class Link {
  /** (optional) Link href */
  @Prop() href: string;
  /** (optional) Disabled link */
  @Prop() disabled?: boolean = false;
  /** (optional) Block link */
  @Prop() block?: boolean = false;
  /** (optional) Link underline */
  // @Prop() underline?: boolean = true;
  /** (optional) Link open a new tag */
  @Prop() target?: string = '_self';
  /** (optional) Link variant */
  @Prop() variant?: string = '';
  /** (optional) Icon size */
  @Prop() iconSize?: number = 24;
  /** (optional) Icon only */
  @Prop() icon?: string;
  /** (optional) Icon title for accessibility */
  @Prop() iconAccessibilityTitle?: string;
  /** (optional) Injected CSS styles */
  @Prop() styles?: string;

  render() {
    return (
      <Host>
        {this.styles && <style>{this.styles}</style>}
        <a
          class={this.getCssClassMap()}
          href={this.disabled ? 'javascript:void(0)' : this.href}
          {...(!this.disabled ? { target: this.target } : {})}
          aria-disabled={this.disabled}
        >
          <span class="link__wrapper">
            <slot />
            {this.icon && this.icon !== '' && (
              <scale-icon
                accessibilityTitle={this.iconAccessibilityTitle}
                path={this.icon}
                size={this.iconSize}
              />
            )}
          </span>
        </a>
      </Host>
    );
  }

  getCssClassMap() {
    return classNames(
      'link',
      this.variant && `$link--variant-${this.variant}`,
      this.disabled && 'link--disabled',
      this.block && 'link--block'
    );
  }
}
