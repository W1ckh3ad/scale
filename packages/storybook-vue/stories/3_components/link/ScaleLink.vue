<template>
    <scale-link
        :block="block"
        :custom-lass="customClass"
        :disabled="JSON.stringify(disabled)"
        :href="href"
        :icon="icon"
        :icon-size="iconSize"
        :target="target"
        :underline="JSON.stringify(underline)"
        :variant="variant"
    >
        <slot></slot>
    </scale-link>
</template>

<script>
export default {
    props: {
        block: {type: Boolean, default: false },
        customClass: String,
        disabled: {type: Boolean, default: false},
        href: String,
        icon: String,
        iconSize: {type: Number, default: 16},
        target: {type: String, default: '_self'},
        underline: Boolean,
        variant: String
    }

}
</script>
