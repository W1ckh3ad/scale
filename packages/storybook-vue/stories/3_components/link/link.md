# Link

Links are used as navigational elements. They may appear on their own, within a sentence or paragraph or directly following content.

## General

Links provide a lightweight option for navigation. Like other interactive elements, too many links will clutter a page and make it difficult for users to identify their next steps. This holds especially true for inline links, which should be used sparingly.

### When to Use

Use links when you want users to:

- Navigate to a different page within the application
- Navigate to an entirely different site
- Jump to an element on the same page
- Link to emails or phone numbers

### When Not to Use:

Whenever you want users to perform actions that will either change data or manipulate the way it is displayed, change a state or trigger an action, use a button instead.

## Variants

### Text Only

We recommend links with three words or less. Because links take users to a new location, it is important that their labels accurately reflect the content users will find there. Therefore, use meaningful labels and avoid terms like “click here” or the web address itself.

Links need to be clear enough for the user to understand, but short enough to avoid that the text wraps onto the next line — unless it is an inline link.

![Image Name](/assets/3_components/links/image-20200810095435628.png)

### Text & Icons

Icons can be added to links in order to highlight certain categories, e.g. links to external sites, to documents, to downloads etc. Make sure to use link icons consistently throughout your product.

![Image Name](/assets/3_components/links/image-20200810095443120.png)

## Sizes

Choose the size according to the body copy you are using.

### Large

20px

![Image Name](/assets/3_components/links/image-20200810095449927.png)

### Small

16px

![Image Name](/assets/3_components/links/image-20200810095443120.png)

## Code
