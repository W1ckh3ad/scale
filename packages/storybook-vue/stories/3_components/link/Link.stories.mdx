import { Meta, Story, ArgsTable, Canvas } from '@storybook/addon-docs/blocks';
import ScaleLink from './ScaleLink.vue';
export const ICON = 'M12.65 3l1.5 1.5H4.5V18c0 .85.65 1.5 1.5 1.5h12c.85 0 1.5-.65 1.5-1.5V9.85l1.5 1.5V18c0 1.65-1.35 3-3 3H6c-1.65 0-3-1.35-3-3V3zM22.7 1.35l-1.65 7.9-2.6-2.6-7.55 7.55c-.15.15-.35.2-.55.2s-.4-.05-.55-.2c-.3-.3-.3-.75 0-1.05l7.55-7.55-2.6-2.6 7.95-1.65z';

<Meta
    title="Components/Link"
    component={ScaleLink}
    argTypes={{
        label: {
            control: {type: 'text'}
        },
        styles: {
            control: {disable: true}
        }
    }}
/>

export const Template = (args, {argTypes}) => ({
    components: {ScaleLink},
    props: {
        label: String,
        ...ScaleLink.props,
    },
    template: `
        <scale-link
        :block="block"
        :custom-lass="customClass"
        :disabled="disabled"
        :href="href"
        :icon="icon"
        :icon-size="iconSize"
        :icon-accessibility-title="iconAccessibilityTitle"
        :target="target"
        :variant="variant"
    >
    {{ label }}
    </scale-link>
    `
})

<div style={{display: 'inline-flex', alignItems: 'center', justifyContent: 'space-between', width: '100%'}}>
    <h1>Link</h1>
    <img src="assets/aa.png" alt="Accessible AA" />
</div>

## Standard

<Canvas withSource="none">
    <Story
        name="Standard"
        args={{
            label: 'A link',
            href: '#top'
        }}>
        {Template.bind({})}
    </Story>
</Canvas>

<ArgsTable story="Standard"/>

## Disabled

<Canvas withSource="none">
    <Story
        name="Disabled"
        args={{
            label: 'A link, disabled',
            disabled: true,
            href: '#top'
        }}
    >
    {Template.bind({})}
    </Story>
</Canvas>

```html
<scale-link disabled href="#top">A link, disabled</scale-link>
```


## With Icon

<Canvas withSource="none">
    <Story
        name="With Icon"
        args={{
            label: 'A link, with an icon',
            icon: ICON,
            href: '#top',
            iconAccessibilityTitle: 'open link in new tab'
        }}>
        {Template.bind({})}
    </Story>
</Canvas>

```html
<scale-link icon="SVG-PATH" href="#top">A link, with an icon</scale-link>
```
