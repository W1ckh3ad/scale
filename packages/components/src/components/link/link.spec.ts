import { newSpecPage } from '@stencil/core/testing';
import { Link } from './link';
import { styles } from './link.styles';
import jss from 'jss';

describe('Link', () => {
  let element;
  let stylesheet;
  beforeEach(async () => {
    element = new Link();
    stylesheet = element.stylesheet = jss.createStyleSheet(styles as any);
  });

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [Link],
      html: `<scale-link>default</scale-link>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should contain target="_blank" when openNewTab is set true', async () => {
    const page = await newSpecPage({
      components: [Link],
      html: `<scale-link href="http://example.com" open-new-tab=true>Label</scale-link>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should have link href value when href value is set', async () => {
    const page = await newSpecPage({
      components: [Link],
      html: `<scale-link href="http://example.com">Label</scale-link>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should handle css classes', () => {
    element.customClass = 'custom';
    expect(element.getCssClassMap()).toContain('custom');

    element.variant = 'primary';
    stylesheet.addRule('link--variant-primary', {});
    expect(element.getCssClassMap()).toContain(
      stylesheet.classes['link--variant-primary']
    );

    element.disabled = true;
    stylesheet.addRule('link--disabled', {});
    expect(element.getCssClassMap()).toContain(
      stylesheet.classes['link--disabled']
    );

    element.underline = true;
    stylesheet.addRule('link--underline', {});
    expect(element.getCssClassMap()).toContain(
      stylesheet.classes['link--underline']
    );
  });
});
