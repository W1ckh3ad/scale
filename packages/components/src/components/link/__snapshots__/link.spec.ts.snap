// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should contain target="_blank" when openNewTab is set true 1`] = `
<scale-link href="http://example.com" open-new-tab="true">
  <mock:shadow-root>
    <style>
      .link-0-0-5 {
  cursor: pointer;
  outline: none;
  text-decoration: none;
  color: #007DB3;
}
.link-0-0-5:hover.link-0-0-5:not(.link--disabled-0-0-8) {
  color: #007DB3;
  border-bottom: 1px solid currentColor;
}
.link-0-0-5:focus:not(.link--disabled-0-0-8) {
  border-bottom: 2px solid currentColor;
}
.link-0-0-5:active.link-0-0-5:not(.link--disabled-0-0-8) {
  color: #0070A0;
}
.link-0-0-5:visited.link-0-0-5:not(.link--disabled-0-0-8) {
  color: #0070A0;
}
.link__wrapper-0-0-6 {
  display: inline-flex;
  align-items: center;
}
.link__wrapper-0-0-6 > *:not(:first-child) {
  margin-inline-start: 0.25rem;
}
.link--block-0-0-7 {
  display: block;
}
.link--disabled-0-0-8 {
  cursor: not-allowed;
  color: #7C7C7C;
}
    </style>
    <a class="link-0-0-5" href="http://example.com" tabindex="0" target="_self">
      <span class="link__wrapper-0-0-6">
        <slot></slot>
      </span>
    </a>
  </mock:shadow-root>
  Label
</scale-link>
`;

exports[`Link should have link href value when href value is set 1`] = `
<scale-link href="http://example.com">
  <mock:shadow-root>
    <style>
      .link-0-0-5 {
  cursor: pointer;
  outline: none;
  text-decoration: none;
  color: #007DB3;
}
.link-0-0-5:hover.link-0-0-5:not(.link--disabled-0-0-8) {
  color: #007DB3;
  border-bottom: 1px solid currentColor;
}
.link-0-0-5:focus:not(.link--disabled-0-0-8) {
  border-bottom: 2px solid currentColor;
}
.link-0-0-5:active.link-0-0-5:not(.link--disabled-0-0-8) {
  color: #0070A0;
}
.link-0-0-5:visited.link-0-0-5:not(.link--disabled-0-0-8) {
  color: #0070A0;
}
.link__wrapper-0-0-6 {
  display: inline-flex;
  align-items: center;
}
.link__wrapper-0-0-6 > *:not(:first-child) {
  margin-inline-start: 0.25rem;
}
.link--block-0-0-7 {
  display: block;
}
.link--disabled-0-0-8 {
  cursor: not-allowed;
  color: #7C7C7C;
}
    </style>
    <a class="link-0-0-5" href="http://example.com" tabindex="0" target="_self">
      <span class="link__wrapper-0-0-6">
        <slot></slot>
      </span>
    </a>
  </mock:shadow-root>
  Label
</scale-link>
`;

exports[`Link should match snapshot 1`] = `
<scale-link>
  <mock:shadow-root>
    <style>
      .link-0-0-5 {
  cursor: pointer;
  outline: none;
  text-decoration: none;
  color: #007DB3;
}
.link-0-0-5:hover.link-0-0-5:not(.link--disabled-0-0-8) {
  color: #007DB3;
  border-bottom: 1px solid currentColor;
}
.link-0-0-5:focus:not(.link--disabled-0-0-8) {
  border-bottom: 2px solid currentColor;
}
.link-0-0-5:active.link-0-0-5:not(.link--disabled-0-0-8) {
  color: #0070A0;
}
.link-0-0-5:visited.link-0-0-5:not(.link--disabled-0-0-8) {
  color: #0070A0;
}
.link__wrapper-0-0-6 {
  display: inline-flex;
  align-items: center;
}
.link__wrapper-0-0-6 > *:not(:first-child) {
  margin-inline-start: 0.25rem;
}
.link--block-0-0-7 {
  display: block;
}
.link--disabled-0-0-8 {
  cursor: not-allowed;
  color: #7C7C7C;
}
    </style>
    <a class="link-0-0-5" tabindex="0" target="_self">
      <span class="link__wrapper-0-0-6">
        <slot></slot>
      </span>
    </a>
  </mock:shadow-root>
  default
</scale-link>
`;
